# Databases and resources {.unnumbered}

1. [GenBank](https://www.ncbi.nlm.nih.gov/genbank/)
2. [UniProt](https://www.uniprot.org/)
3. [PubMed](https://pubmed.ncbi.nlm.nih.gov/)
4. [Ensembl](https://www.ensembl.org/)
5. [NCBI Gene](https://www.ncbi.nlm.nih.gov/gene)
6. [STRING](https://string-db.org/)
7. [dbSNP](https://www.ncbi.nlm.nih.gov/snp/)
8. [OMIM](https://www.omim.org/)
9. [PDB (Protein Data Bank)](https://www.rcsb.org/)
10. [KEGG (Kyoto Encyclopedia of Genes and Genomes)](https://www.genome.jp/kegg/)
11. [Reactome](https://reactome.org/)
12. [GO (Gene Ontology)](http://geneontology.org/)
13. [HPRD (Human Protein Reference Database)](http://www.hprd.org/)
14. [COSMIC (Catalogue Of Somatic Mutations In Cancer)](https://cancer.sanger.ac.uk/cosmic)
15. [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/)
16. [TCGA (The Cancer Genome Atlas)](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga)
17. [Pfam](https://pfam.xfam.org/)
18. [Rfam](https://rfam.xfam.org/)
19. [InterPro](https://www.ebi.ac.uk/interpro/)
20. [GTEx (Genotype-Tissue Expression Project)](https://gtexportal.org/home/)
21. [HGNC (HUGO Gene Nomenclature Committee)](https://www.genenames.org/)
22. [FlyBase](https://flybase.org/)
23. [WormBase](https://www.wormbase.org/)
24. [TAIR (The Arabidopsis Information Resource)](https://www.arabidopsis.org/)
25. [Mouse Genome Informatics (MGI)](http://www.informatics.jax.org/)
26. [RGD (Rat Genome Database)](https://rgd.mcw.edu/)
27. [PharmGKB](https://www.pharmgkb.org/)
28. [DrugBank](https://go.drugbank.com/)
29. [PubChem](https://pubchem.ncbi.nlm.nih.gov/)
30. [dbGaP (Database of Genotypes and Phenotypes)](https://www.ncbi.nlm.nih.gov/gap)
31. [dbCAN (Carbohydrate-Active enZYmes Database)](http://bcb.unl.edu/dbCAN/)
32. [CATH (Class, Architecture, Topology, Homology)](http://www.cathdb.info/)
33. [MEROPS (Peptidase Database)](https://www.ebi.ac.uk/merops/)
34. [BioGRID](https://thebiogrid.org/)
35. [IntAct](https://www.ebi.ac.uk/intact/)
36. [IUPHAR/BPS Guide to Pharmacology](https://www.guidetopharmacology.org/)
37. [GenAtlas](https://genatlas.medecine.univ-paris5.fr/)
38. [miRBase](http://www.mirbase.org/)
39. [lncRNAdb](http://www.lncrnadb.org/)
40. [DGV (Database of Genomic Variants)](http://dgv.tcag.ca/)
41. [UCSC Genome Browser](https://genome.ucsc.edu/)
42. [ExPASy (Expert Protein Analysis System)](https://web.expasy.org/)
43. [Swiss-Model](https://swissmodel.expasy.org/)
44. [Pfam](https://pfam.xfam.org/)
45. [SUPERFAMILY](http://supfam.cs.bris.ac.uk/)
46. [EMBL-EBI InterProScan](https://www.ebi.ac.uk/interpro/interproscan.html)
47. [WikiPathways](https://www.wikipathways.org/)
48. [ChEMBL](https://www.ebi.ac.uk/chembl/)
49. [BioCyc](https://biocyc.org/)
50. [dbVar](https://www.ncbi.nlm.nih.gov/dbvar/)




## Knowledege data bases  {.unnumbered}

## Do it yourself:   {.unnumbered}


